# SENTIMENT ANALYSIS

# INTRODUCTION

This is a list of 1,000 hotels and their reviews provided by Datafiniti's Business Database.


The dataset includes hotel location, name, rating, review data, title, username, and more.

# About Dataset

ALTHOUGH ORIGINAL DATASET CONTAINS MANY COLUMNS WE ONLY REQUIRE THESE MANY



1>address--------> Address of the Hotel  ('5620 calle real', '5th and san carlos po box 3574',
                                         '167 w main st','3101 coliseum dr', '7886 main street',
                                         '3505 s main st'...)   




   
2>city-----------> Cities                ('goleta', 'carmel by the sea', 'lexington', 'russell springs',
                                         'hopkinsville', 'carpinteria', 'redwood city', 'new orleans',
                                         'estes park', 'rangely', 'longmont', 'colorado springs', 'macomb'.....)






                        
3>name-----------> Name of Hotels        ('best western plus south coast inn',
                                         "best western carmel's town house lodge",
                                         '21c museum hotel lexington', ...)





     
      
4>reviews.rating--> Ratings given        (1,2,3,4,5)        
   
    
    
    
    
    
    
    
reviews.text------> Reviews given                   
            

# IMPORTING LIBRARIES

In [135]:
import pandas as pd
import numpy as np
import string
import nltk
nltk.download('punkt')
nltk.download('stopwords')
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\PRANAV\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\PRANAV\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [136]:
data=pd.read_csv(r"C:\Users\PRANAV\Downloads\archive (16)\Datafiniti_Hotel_Reviews_Jun19.csv")         # reading dataset

In [137]:
data.head()

id             dateAdded           dateUpdated  \
0  AWE2FvX5RxPSIh2RscTK  2018-01-18T18:43:12Z  2019-05-20T23:55:47Z   
1  AVwcj_OhkufWRAb5wi9T  2016-11-06T20:21:05Z  2019-05-20T23:31:56Z   
2  AVwcj_OhkufWRAb5wi9T  2016-11-06T20:21:05Z  2019-05-20T23:31:56Z   
3  AVwcj_OhkufWRAb5wi9T  2016-11-06T20:21:05Z  2019-05-20T23:31:56Z   
4  AVwcj_OhkufWRAb5wi9T  2016-11-06T20:21:05Z  2019-05-20T23:31:56Z   

                          address  \
0                 5620 Calle Real   
1  5th And San Carlos PO Box 3574   
2  5th And San Carlos PO Box 3574   
3  5th And San Carlos PO Box 3574   
4  5th And San Carlos PO Box 3574   

                                          categories  \
0  Hotels,Hotels and motels,Hotel and motel mgmt....   
1                               Hotels,Lodging,Hotel   
2                               Hotels,Lodging,Hotel   
3                               Hotels,Lodging,Hotel   
4                               Hotels,Lodging,Hotel   

               primaryCategories               city country  \
0  Accommodation & Food Services             Goleta      US   
1  Accommodation & Food Services  Carmel by the Sea      US   
2  Accommodation & Food Services  Carmel by the Sea      US   
3  Accommodation & Food Services  Carmel by the Sea      US   
4  Accommodation & Food Services  Carmel by the Sea      US   

                                                keys  latitude  ...  \
0             us/ca/goleta/5620callereal/-1127060008  34.44178  ...   
1  us/ca/carmelbythesea/5thandsancarlospobox3574/...  36.55722  ...   
2  us/ca/carmelbythesea/5thandsancarlospobox3574/...  36.55722  ...   
3  us/ca/carmelbythesea/5thandsancarlospobox3574/...  36.55722  ...   
4  us/ca/carmelbythesea/5thandsancarlospobox3574/...  36.55722  ...   

                            reviews.dateSeen reviews.rating  \
0                       2018-01-03T00:00:00Z              3   
1                       2016-10-09T00:00:00Z              4   
2                       2016-10-09T00:00:00Z              3   
3                       2016-10-31T00:00:00Z              5   
4  2016-10-09T00:00:00Z,2016-03-27T00:00:00Z              2   

                                  reviews.sourceURLs  \
0  https://www.tripadvisor.com/Hotel_Review-g3243...   
1  https://www.tripadvisor.com/Hotel_Review-g3217...   
2  https://www.tripadvisor.com/Hotel_Review-g3217...   
3  https://www.tripadvisor.com/Hotel_Review-g3217...   
4  https://www.tripadvisor.com/Hotel_Review-g3217...   

                                        reviews.text  \
0  This hotel was nice and quiet. Did not know, t...   
1  We stayed in the king suite with the separatio...   
2  Parking was horrible, somebody ran into my ren...   
3  Not cheap but excellent location. Price is som...   
4  If you get the room that they advertised on th...   

                                       reviews.title  reviews.userCity  \
0                            Best Western Plus Hotel          San Jose   
1  Clean rooms at solid rates in the heart of Carmel     San Francisco   
2                                           Business   Prescott Valley   
3                                          Very good          Guaynabo   
4                       Low chance to come back here              Reno   

  reviews.userProvince  reviews.username  \
0         UnitedStates      tatsurok2018   
1                   CA         STEPHEN N   
2                   AZ         15Deborah   
3                   PR        Wilfredo M   
4                   NV             Luc D   

                                          sourceURLs  \
0  https://www.tripadvisor.com/Hotel_Review-g3243...   
1  http://www.tripadvisor.com/Hotel_Review-g32172...   
2  http://www.tripadvisor.com/Hotel_Review-g32172...   
3  http://www.tripadvisor.com/Hotel_Review-g32172...   
4  http://www.tripadvisor.com/Hotel_Review-g32172...   

                                            websites  
0  https://www.bestwestern.com/en_US/book/hotel-r...  
1                 

In [138]:
data.shape           # Data contains 10,000 records and 26 columns

(10000, 26)

In [139]:
data.info()          

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 26 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    10000 non-null  object 
 1   dateAdded             10000 non-null  object 
 2   dateUpdated           10000 non-null  object 
 3   address               10000 non-null  object 
 4   categories            10000 non-null  object 
 5   primaryCategories     10000 non-null  object 
 6   city                  10000 non-null  object 
 7   country               10000 non-null  object 
 8   keys                  10000 non-null  object 
 9   latitude              10000 non-null  float64
 10  longitude             10000 non-null  float64
 11  name                  10000 non-null  object 
 12  postalCode            10000 non-null  object 
 13  province              10000 non-null  object 
 14  reviews.date          10000 non-null  object 
 15  reviews.dateAdded   

In [140]:
data=data.loc[::,['address','city','name','reviews.text','reviews.rating']]           # Selecting only those columns 
                                                                                      #  which are needed

In [141]:
data

address               city  \
0                    5620 Calle Real             Goleta   
1     5th And San Carlos PO Box 3574  Carmel by the Sea   
2     5th And San Carlos PO Box 3574  Carmel by the Sea   
3     5th And San Carlos PO Box 3574  Carmel by the Sea   
4     5th And San Carlos PO Box 3574  Carmel by the Sea   
...                              ...                ...   
9995                3101 Coliseum Dr            Hampton   
9996                3101 Coliseum Dr            Hampton   
9997                3101 Coliseum Dr            Hampton   
9998                7886 Main Street             Hunter   
9999                  3505 S Main St            Lindale   

                                        name  \
0          Best Western Plus South Coast Inn   
1     Best Western Carmel's Town House Lodge   
2     Best Western Carmel's Town House Lodge   
3     Best Western Carmel's Town House Lodge   
4     Best Western Carmel's Town House Lodge   
...                                      ...   
9995        Hampton Inn Hampton-newport News   
9996        Hampton Inn Hampton-newport News   
9997        Hampton Inn Hampton-newport News   
9998                         Roseberry's Inn   
9999               Hampton Inn-lindale/tyler   

                                           reviews.text  reviews.rating  
0     This hotel was nice and quiet. Did not know, t...               3  
1     We stayed in the king suite with the separatio...               4  
2     Parking was horrible, somebody ran into my ren...               3  
3     Not cheap but excellent location. Price is som...               5  
4     If you get the room that they advertised on th...               2  
...                                                 ...             ...  
9995  My friends and I took a trip to Hampton for th...               4  
9996  from check in to departure, staff is friendly,...               5  
9997  This Hampton is located on a quiet street acro...               5  
9998  Awesome wings (my favorite was garlic parmesan...               5  
9999  Clean facility just off freeway ..... staff fr...               4  

[10000 rows x 5 columns]

# CONVERTING COLUMNS INTO LOWER CASE

In [142]:
data.address=data.address.str.lower()
data.city=data.city.str.lower()
data.name=data.name.str.lower()
data['reviews.text']=data['reviews.text'].str.lower()

In [143]:
data.head()

address               city  \
0                 5620 calle real             goleta   
1  5th and san carlos po box 3574  carmel by the sea   
2  5th and san carlos po box 3574  carmel by the sea   
3  5th and san carlos po box 3574  carmel by the sea   
4  5th and san carlos po box 3574  carmel by the sea   

                                     name  \
0       best western plus south coast inn   
1  best western carmel's town house lodge   
2  best western carmel's town house lodge   
3  best western carmel's town house lodge   
4  best western carmel's town house lodge   

                                        reviews.text  reviews.rating  
0  this hotel was nice and quiet. did not know, t...               3  
1  we stayed in the king suite with the separatio...               4  
2  parking was horrible, somebody ran into my ren...               3  
3  not cheap but excellent location. price is som...               5  
4  if you get the room that they advertised on th...               2

In [144]:
abc=list(nltk.corpus.stopwords.words('english'))            # List of 'stopwords'

In [145]:
abc

['i',
 'me',
 'my',
 'myself',
 'we',
 'our',
 'ours',
 'ourselves',
 'you',
 "you're",
 "you've",
 "you'll",
 "you'd",
 'your',
 'yours',
 'yourself',
 'yourselves',
 'he',
 'him',
 'his',
 'himself',
 'she',
 "she's",
 'her',
 'hers',
 'herself',
 'it',
 "it's",
 'its',
 'itself',
 'they',
 'them',
 'their',
 'theirs',
 'themselves',
 'what',
 'which',
 'who',
 'whom',
 'this',
 'that',
 "that'll",
 'these',
 'those',
 'am',
 'is',
 'are',
 'was',
 'were',
 'be',
 'been',
 'being',
 'have',
 'has',
 'had',
 'having',
 'do',
 'does',
 'did',
 'doing',
 'a',
 'an',
 'the',
 'and',
 'but',
 'if',
 'or',
 'because',
 'as',
 'until',
 'while',
 'of',
 'at',
 'by',
 'for',
 'with',
 'about',
 'against',
 'between',
 'into',
 'through',
 'during',
 'before',
 'after',
 'above',
 'below',
 'to',
 'from',
 'up',
 'down',
 'in',
 'out',
 'on',
 'off',
 'over',
 'under',
 'again',
 'further',
 'then',
 'once',
 'here',
 'there',
 'when',
 'where',
 'why',
 'how',
 'all',
 'any',
 'both',
 'each

In [146]:
import string

In [147]:
string.punctuation

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

CREATING USER DEFINED FUNCTION TO REMOVE DUPLICATES AND STOPWORDS

In [148]:
def text_process(mess):            ### creating a function
    """                                                        ## a docstring
    1. remove the punctuation
    2. remove the stopwords
    3. return the list of clean textwords
    
    """
    nopunc = [char for char in mess if char not in string.punctuation]
    nopunc = "".join(nopunc)
    
    return [ word for word in nopunc.split() if word not  in abc]

In [149]:
data['reviews.text'].apply(text_process)              # Removing stopwords and punctuations

0       [hotel, nice, quiet, know, train, track, near,...
1       [stayed, king, suite, separation, bedroom, liv...
2       [parking, horrible, somebody, ran, rental, car...
3       [cheap, excellent, location, price, somewhat, ...
4       [get, room, advertised, website, paid, may, lu...
                              ...                        
9995    [friends, took, trip, hampton, basslights, sho...
9996    [check, departure, staff, friendly, profession...
9997    [hampton, located, quiet, street, across, hosp...
9998    [awesome, wings, favorite, garlic, parmesan, g...
9999    [clean, facility, freeway, staff, friendly, ef...
Name: reviews.text, Length: 10000, dtype: object

In [709]:
from sklearn.feature_extraction.text import CountVectorizer                

In [710]:
bow=CountVectorizer(analyzer=text_process).fit(data['reviews.text'])
bow.vocabulary_                                                                        # frequency of words

{'hotel': 14131,
 'nice': 19195,
 'quiet': 22611,
 'know': 16119,
 'train': 29060,
 'track': 29030,
 'near': 18979,
 'passed': 20810,
 'stay': 26955,
 'best': 4164,
 'western': 31087,
 'changed': 5882,
 'classification': 6340,
 'plus': 21572,
 'category': 5663,
 'stayed': 26979,
 'king': 16042,
 'suite': 27504,
 'separation': 25124,
 'bedroom': 3977,
 'living': 16820,
 'space': 26474,
 'sofa': 26190,
 'bed': 3948,
 'wasnt': 30769,
 'good': 12721,
 'back': 3490,
 'discomfort': 8901,
 'day': 8060,
 'left': 16492,
 'three': 28490,
 'night': 19270,
 'room': 24192,
 'clean': 6355,
 'comfortable': 6716,
 'located': 16890,
 'within': 31393,
 'walking': 30623,
 'distance': 9040,
 'places': 21407,
 'want': 30688,
 'parking': 20697,
 'horrible': 14052,
 'somebody': 26248,
 'ran': 22764,
 'rental': 23479,
 'car': 5471,
 'staying': 26995,
 'didnt': 8673,
 'get': 12523,
 'try': 29356,
 'breakfast': 4778,
 'business': 5164,
 'restaurant': 23697,
 'opened': 20073,
 'late': 16345,
 'world': 31530,
 'e

In [711]:
tdm=bow.transform(data['reviews.text'])                            # Term Document Matrix

In [712]:
from sklearn.model_selection import train_test_split

In [713]:
train_x,test_x,train_y,test_y=train_test_split(tdm,data['reviews.rating'],test_size=0.2)

In [714]:
from sklearn.tree import DecisionTreeClassifier
dt = DecisionTreeClassifier()

In [715]:
dt.fit(train_x, train_y)

DecisionTreeClassifier()

In [716]:
pred_test=dt.predict(test_x)

In [717]:
from sklearn.metrics import *

In [718]:
tab=confusion_matrix(test_y,pred_test)
tab

array([[ 29,  21,  28,  16,  23],
       [ 18,  19,  25,  28,  28],
       [ 25,  21,  54,  77,  64],
       [ 11,  14,  57, 231, 291],
       [ 16,  11,  40, 224, 629]], dtype=int64)

In [719]:
accuracy_dt=accuracy_score(test_y,pred_test)*100
accuracy_dt

48.1

In [720]:
print(classification_report(test_y,pred_test))

              precision    recall  f1-score   support

           1       0.29      0.25      0.27       117
           2       0.22      0.16      0.19       118
           3       0.26      0.22      0.24       241
           4       0.40      0.38      0.39       604
           5       0.61      0.68      0.64       920

    accuracy                           0.48      2000
   macro avg       0.36      0.34      0.35      2000
weighted avg       0.46      0.48      0.47      2000



# CHANGING CATEGORIES OF REVIEWS FROM 5 TO 3

In [150]:
data['reviews.rating'].replace({2:1,3:2,4:3,5:3},inplace=True)          # Now only 3 Categories are there

In [774]:
from sklearn.feature_extraction.text import CountVectorizer


In [775]:
bow=CountVectorizer(analyzer=text_process).fit(data['reviews.text'])

In [776]:
tdm=bow.transform(data['reviews.text'])

In [777]:
from sklearn.model_selection import train_test_split

train_x,test_x,train_y,test_y=train_test_split(tdm,data['reviews.rating'],test_size=0.2)


In [650]:
from sklearn.tree import DecisionTreeClassifier
dt = DecisionTreeClassifier()

In [651]:
dt.fit(train_x,train_y)

DecisionTreeClassifier()

In [438]:
pred_test=dt.predict(test_x)

In [439]:
from sklearn.metrics import *

In [440]:
tab=confusion_matrix(test_y,pred_test)
tab

array([[ 104,   38,   72],
       [  42,   50,  146],
       [  53,   89, 1406]], dtype=int64)

In [441]:

accuracy_dt=accuracy_score(test_y,pred_test)*100
accuracy_dt

78.0

In [442]:
print(classification_report(test_y,pred_test))

              precision    recall  f1-score   support

           1       0.52      0.49      0.50       214
           2       0.28      0.21      0.24       238
           3       0.87      0.91      0.89      1548

    accuracy                           0.78      2000
   macro avg       0.56      0.53      0.54      2000
weighted avg       0.76      0.78      0.77      2000



SO IT IS GIVING BETTER RESULTS FOR 3 CATEGORIES AND COULD BE SEEN THAT CLASS IMBALANCE PROBLEM IS PRESENT

# HYPERPARAMETER TUNING

HYPERPARMETERS-----------> THESE ARE THOSE PARAMETERS WHICH ARE NOT PART ORIGINAL DATA AND USER CAN CONTROL IT

FOR EX--> CRITERIA ,  MAX_DEPTH , MIN_SAMPLE_SPLIT


SO WE NEED TO FIND THE BEST COMBINATIION OF THSES , TO GET THE OPTIMUM MODEL

In [519]:
from sklearn.model_selection import GridSearchCV

search_dict = {"criterion": ['gini','entropy'],
              "max_depth": range(6,10),
              "min_samples_split": (25, 50, 75,100,150) }

In [520]:
from sklearn.tree import DecisionTreeClassifier
dt = DecisionTreeClassifier()

dt_GridSearchCV = GridSearchCV(dt, param_grid= search_dict) 
# this will build build multiple models and select the best model

In [521]:
dt_GridSearchCV.fit(train_x, train_y)


GridSearchCV(estimator=DecisionTreeClassifier(),
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': range(6, 10),
                         'min_samples_split': (25, 50, 75, 100, 150)})

In [523]:
dt_GridSearchCV.best_params_

{'criterion': 'gini', 'max_depth': 8, 'min_samples_split': 25}

SO NOW WE WILL BE BUILDING MODEL WITH THIS COMBINATION ONLY

# TDM

In [783]:
data['reviews.text'].apply(text_process)

0       [hotel, nice, quiet, know, train, track, near,...
1       [stayed, king, suite, separation, bedroom, liv...
2       [parking, horrible, somebody, ran, rental, car...
3       [cheap, excellent, location, price, somewhat, ...
4       [get, room, advertised, website, paid, may, lu...
                              ...                        
9995    [friends, took, trip, hampton, basslights, sho...
9996    [check, departure, staff, friendly, profession...
9997    [hampton, located, quiet, street, across, hosp...
9998    [awesome, wings, favorite, garlic, parmesan, g...
9999    [clean, facility, freeway, staff, friendly, ef...
Name: reviews.text, Length: 10000, dtype: object

In [784]:
from sklearn.feature_extraction.text import CountVectorizer

In [785]:
bow=CountVectorizer(analyzer=text_process).fit(data['reviews.text'])

In [786]:
bow

CountVectorizer(analyzer=<function text_process at 0x0000018AB5BED480>)

In [787]:
tdm=bow.transform(data['reviews.text'])

In [788]:
from sklearn.model_selection import train_test_split

In [789]:
train_x,test_x,train_y,test_y=train_test_split(tdm,data['reviews.rating'],test_size=0.2)

In [790]:
from sklearn.tree import DecisionTreeClassifier
dt = DecisionTreeClassifier(criterion= 'gini',max_depth=8,min_samples_split=25,)

In [791]:
dt.fit(train_x, train_y)

DecisionTreeClassifier(max_depth=8, min_samples_split=25)

In [792]:
pred_test = dt.predict(test_x)


In [793]:
from sklearn.metrics import *

In [794]:
tab=confusion_matrix(test_y,pred_test)

In [795]:
tab

array([[  80,   12,  128],
       [  29,   18,  185],
       [  32,   11, 1505]], dtype=int64)

In [796]:
accuracy_tdm=accuracy_score(test_y,pred_test)*100
accuracy_tdm

80.15

In [797]:
print(classification_report(test_y,pred_test))

              precision    recall  f1-score   support

           1       0.57      0.36      0.44       220
           2       0.44      0.08      0.13       232
           3       0.83      0.97      0.89      1548

    accuracy                           0.80      2000
   macro avg       0.61      0.47      0.49      2000
weighted avg       0.75      0.80      0.76      2000



# TF-IDF

TF-IDF ---------> TERM FREQUENCY INVERSE DOCUMENT FREQUENCY



IT MAY HAPPEN THAN THE FREQUENCY OF A WORD MAY BE HIGH AND IT IS ALSO NOT A STOP WORD BUT 




STILL IT IS NOT ADDING ANY MEANING.SO FOR THESE SCENARIOS WE USE TF-IDF

In [803]:
data['reviews.text'].apply(text_process)

0       [hotel, nice, quiet, know, train, track, near,...
1       [stayed, king, suite, separation, bedroom, liv...
2       [parking, horrible, somebody, ran, rental, car...
3       [cheap, excellent, location, price, somewhat, ...
4       [get, room, advertised, website, paid, may, lu...
                              ...                        
9995    [friends, took, trip, hampton, basslights, sho...
9996    [check, departure, staff, friendly, profession...
9997    [hampton, located, quiet, street, across, hosp...
9998    [awesome, wings, favorite, garlic, parmesan, g...
9999    [clean, facility, freeway, staff, friendly, ef...
Name: reviews.text, Length: 10000, dtype: object

In [804]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [805]:
tfidf=TfidfVectorizer()

In [806]:
tfidf_metrics=tfidf.fit_transform(data['reviews.text'])

In [807]:
tfidf_metrics

<10000x22112 sparse matrix of type '<class 'numpy.float64'>'
	with 763158 stored elements in Compressed Sparse Row format>

In [808]:
from sklearn.model_selection import train_test_split

In [809]:
train_x,test_x,train_y,test_y=train_test_split(tfidf_metrics,data['reviews.rating'],test_size=0.2)

In [810]:
from sklearn.tree import DecisionTreeClassifier
dt = DecisionTreeClassifier(criterion= 'gini', max_depth= 8, min_samples_split= 25)

In [811]:
dt.fit(train_x,train_y)

DecisionTreeClassifier(max_depth=8, min_samples_split=25)

In [812]:
pred_test=dt.predict(test_x)

In [813]:
from sklearn.metrics import *

In [814]:
tab=confusion_matrix(test_y,pred_test)
tab

array([[  51,   17,  154],
       [  23,   18,  184],
       [  13,   24, 1516]], dtype=int64)

In [815]:
accuracy_tf_idf=accuracy_score(test_y,pred_test)*100
accuracy_tf_idf

79.25

In [816]:
print(classification_report(test_y,pred_test))

              precision    recall  f1-score   support

           1       0.59      0.23      0.33       222
           2       0.31      0.08      0.13       225
           3       0.82      0.98      0.89      1553

    accuracy                           0.79      2000
   macro avg       0.57      0.43      0.45      2000
weighted avg       0.73      0.79      0.74      2000



SO TF-IDF and TDM are giving same results



accuracy_tdm= 80.15%




accuracy_tf_idf= 79.25%


# NAIVE BAYES

In [379]:
data['reviews.text'].apply(text_process)

0       [hotel, nice, quiet, know, train, track, near,...
1       [stayed, king, suite, separation, bedroom, liv...
2       [parking, horrible, somebody, ran, rental, car...
3       [cheap, excellent, location, price, somewhat, ...
4       [get, room, advertised, website, paid, may, lu...
                              ...                        
9995    [friends, took, trip, hampton, basslights, sho...
9996    [check, departure, staff, friendly, profession...
9997    [hampton, located, quiet, street, across, hosp...
9998    [awesome, wings, favorite, garlic, parmesan, g...
9999    [clean, facility, freeway, staff, friendly, ef...
Name: reviews.text, Length: 10000, dtype: object

In [380]:
from sklearn.feature_extraction.text import CountVectorizer

In [381]:
bow=CountVectorizer(analyzer=text_process).fit(data['reviews.text'])

In [382]:

tdm=bow.transform(data['reviews.text'])

In [383]:
from sklearn.model_selection import train_test_split

In [384]:
train_x,test_x,train_y,test_y=train_test_split(tdm,data['reviews.rating'],test_size=0.2)

In [427]:
from sklearn.naive_bayes import MultinomialNB
ml=MultinomialNB(alpha=0.9)

In [428]:
ml.fit(train_x,train_y)

MultinomialNB(alpha=0.9)

In [419]:
pred_test=ml.predict(train_x)

In [420]:
from sklearn.metrics import *


In [397]:
tab=confusion_matrix(train_y,pred_test)
tab

array([[ 785,   15,  109],
       [  37,  491,  402],
       [  46,   32, 6083]], dtype=int64)

In [402]:
accuracy_Naiye_Bayes_train=accuracy_score(train_y,pred_test)*100
accuracy_Naiye_Bayes_train

91.9875

In [401]:
print(classification_report(train_y,pred_test))

              precision    recall  f1-score   support

           1       0.90      0.86      0.88       909
           2       0.91      0.53      0.67       930
           3       0.92      0.99      0.95      6161

    accuracy                           0.92      8000
   macro avg       0.91      0.79      0.84      8000
weighted avg       0.92      0.92      0.91      8000



In [429]:
pred_test=ml.predict(test_x)

In [430]:
from sklearn.metrics import *


In [431]:
tab=confusion_matrix(test_y,pred_test)
tab

array([[ 156,   21,   35],
       [  36,   57,  167],
       [  18,   24, 1486]], dtype=int64)

In [432]:
accuracy_Naiye_Bayes=accuracy_score(test_y,pred_test)*100
accuracy_Naiye_Bayes

84.95

In [433]:
print(classification_report(test_y,pred_test))

              precision    recall  f1-score   support

           1       0.74      0.74      0.74       212
           2       0.56      0.22      0.31       260
           3       0.88      0.97      0.92      1528

    accuracy                           0.85      2000
   macro avg       0.73      0.64      0.66      2000
weighted avg       0.82      0.85      0.83      2000



SO IN BETWEEN Logistic AND NAIYE BAYES , NAIYE BAYES IS GIVING BETTER RESULTS

accuracy_tdm= 80.15%




accuracy_tf_idf= 79.25%






accuracy_Naiye_Bayes=85%

# LSTM

In [21]:
import numpy as np
import pandas as pd
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, LSTM, Dense

In [894]:
data=pd.read_csv(r"C:\Users\PRANAV\Downloads\archive (16)\Datafiniti_Hotel_Reviews_Jun19.csv")
data.head()

id             dateAdded           dateUpdated  \
0  AWE2FvX5RxPSIh2RscTK  2018-01-18T18:43:12Z  2019-05-20T23:55:47Z   
1  AVwcj_OhkufWRAb5wi9T  2016-11-06T20:21:05Z  2019-05-20T23:31:56Z   
2  AVwcj_OhkufWRAb5wi9T  2016-11-06T20:21:05Z  2019-05-20T23:31:56Z   
3  AVwcj_OhkufWRAb5wi9T  2016-11-06T20:21:05Z  2019-05-20T23:31:56Z   
4  AVwcj_OhkufWRAb5wi9T  2016-11-06T20:21:05Z  2019-05-20T23:31:56Z   

                          address  \
0                 5620 Calle Real   
1  5th And San Carlos PO Box 3574   
2  5th And San Carlos PO Box 3574   
3  5th And San Carlos PO Box 3574   
4  5th And San Carlos PO Box 3574   

                                          categories  \
0  Hotels,Hotels and motels,Hotel and motel mgmt....   
1                               Hotels,Lodging,Hotel   
2                               Hotels,Lodging,Hotel   
3                               Hotels,Lodging,Hotel   
4                               Hotels,Lodging,Hotel   

               primaryCategories               city country  \
0  Accommodation & Food Services             Goleta      US   
1  Accommodation & Food Services  Carmel by the Sea      US   
2  Accommodation & Food Services  Carmel by the Sea      US   
3  Accommodation & Food Services  Carmel by the Sea      US   
4  Accommodation & Food Services  Carmel by the Sea      US   

                                                keys  latitude  ...  \
0             us/ca/goleta/5620callereal/-1127060008  34.44178  ...   
1  us/ca/carmelbythesea/5thandsancarlospobox3574/...  36.55722  ...   
2  us/ca/carmelbythesea/5thandsancarlospobox3574/...  36.55722  ...   
3  us/ca/carmelbythesea/5thandsancarlospobox3574/...  36.55722  ...   
4  us/ca/carmelbythesea/5thandsancarlospobox3574/...  36.55722  ...   

                            reviews.dateSeen reviews.rating  \
0                       2018-01-03T00:00:00Z              3   
1                       2016-10-09T00:00:00Z              4   
2                       2016-10-09T00:00:00Z              3   
3                       2016-10-31T00:00:00Z              5   
4  2016-10-09T00:00:00Z,2016-03-27T00:00:00Z              2   

                                  reviews.sourceURLs  \
0  https://www.tripadvisor.com/Hotel_Review-g3243...   
1  https://www.tripadvisor.com/Hotel_Review-g3217...   
2  https://www.tripadvisor.com/Hotel_Review-g3217...   
3  https://www.tripadvisor.com/Hotel_Review-g3217...   
4  https://www.tripadvisor.com/Hotel_Review-g3217...   

                                        reviews.text  \
0  This hotel was nice and quiet. Did not know, t...   
1  We stayed in the king suite with the separatio...   
2  Parking was horrible, somebody ran into my ren...   
3  Not cheap but excellent location. Price is som...   
4  If you get the room that they advertised on th...   

                                       reviews.title  reviews.userCity  \
0                            Best Western Plus Hotel          San Jose   
1  Clean rooms at solid rates in the heart of Carmel     San Francisco   
2                                           Business   Prescott Valley   
3                                          Very good          Guaynabo   
4                       Low chance to come back here              Reno   

  reviews.userProvince  reviews.username  \
0         UnitedStates      tatsurok2018   
1                   CA         STEPHEN N   
2                   AZ         15Deborah   
3                   PR        Wilfredo M   
4                   NV             Luc D   

                                          sourceURLs  \
0  https://www.tripadvisor.com/Hotel_Review-g3243...   
1  http://www.tripadvisor.com/Hotel_Review-g32172...   
2  http://www.tripadvisor.com/Hotel_Review-g32172...   
3  http://www.tripadvisor.com/Hotel_Review-g32172...   
4  http://www.tripadvisor.com/Hotel_Review-g32172...   

                                            websites  
0  https://www.bestwestern.com/en_US/book/hotel-r...  
1                 

In [16]:
data=data.loc[::,['reviews.rating','reviews.text']]
data

reviews.rating                                       reviews.text
0                  3  this hotel was nice and quiet. did not know, t...
1                  4  we stayed in the king suite with the separatio...
2                  3  parking was horrible, somebody ran into my ren...
3                  5  not cheap but excellent location. price is som...
4                  2  if you get the room that they advertised on th...
...              ...                                                ...
9995               4  my friends and i took a trip to hampton for th...
9996               5  from check in to departure, staff is friendly,...
9997               5  this hampton is located on a quiet street acro...
9998               5  awesome wings (my favorite was garlic parmesan...
9999               4  clean facility just off freeway ..... staff fr...

[10000 rows x 2 columns]

In [17]:
data['reviews.text'].apply(text_process)

0       [hotel, nice, quiet, know, train, track, near,...
1       [stayed, king, suite, separation, bedroom, liv...
2       [parking, horrible, somebody, ran, rental, car...
3       [cheap, excellent, location, price, somewhat, ...
4       [get, room, advertised, website, paid, may, lu...
                              ...                        
9995    [friends, took, trip, hampton, basslights, sho...
9996    [check, departure, staff, friendly, profession...
9997    [hampton, located, quiet, street, across, hosp...
9998    [awesome, wings, favorite, garlic, parmesan, g...
9999    [clean, facility, freeway, staff, friendly, ef...
Name: reviews.text, Length: 10000, dtype: object

In [18]:
data['reviews.rating'].replace({1:0,2:0,3:1,4:2,5:2},inplace=True) 

In [19]:
data_x=data.iloc[:,1]
data_y=data.iloc[:,0]
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(data_x,data_y,test_size=0.2)

In [22]:
y_train=to_categorical(y_train)

In [23]:
max_num_words=10000        # max words to be taken from entire corpus
seq_len=150                 # words taken from each document
embedding_size=100          # embedding size of each word

In [24]:
from keras.preprocessing.text import Tokenizer # is used for tokenization assign id(num) to each word
from keras.preprocessing.sequence import pad_sequences # used to bring all samples to same length by adding extra 0

In [25]:
tokenizer=Tokenizer(num_words=max_num_words)
tokenizer.fit_on_texts(data['reviews.text'])
x_train=tokenizer.texts_to_sequences(x_train)        # will convert text to sequences of Ids
x_train=pad_sequences(x_train,maxlen=seq_len)

x_test=tokenizer.texts_to_sequences(x_test)
x_test=pad_sequences(x_test,maxlen=seq_len)

model=Sequential()  # initilize the network
model.add(Embedding(input_dim=max_num_words,
                   input_length=seq_len,
                   output_dim=embedding_size))

In [26]:
model.add(LSTM(100))
model.add(Dense(3,activation='softmax'))# this 3 bcoz u have 3 level in target variable
from tensorflow.keras.optimizers import Adam
adam=Adam(lr=0.0001)
model.compile(optimizer='adam',loss='categorical_crossentropy',metrics=['accuracy'])

In [27]:
model.fit(x_train,y_train,epochs=8,batch_size=32,validation_split=.2)

Epoch 1/8
200/200 [==============================] - 20s 89ms/step - loss: 0.6130 - accuracy: 0.7833 - val_loss: 0.4475 - val_accuracy: 0.8263
Epoch 2/8
200/200 [==============================] - 19s 95ms/step - loss: 0.4014 - accuracy: 0.8427 - val_loss: 0.4404 - val_accuracy: 0.8256
Epoch 3/8
200/200 [==============================] - 19s 96ms/step - loss: 0.3028 - accuracy: 0.8775 - val_loss: 0.4623 - val_accuracy: 0.8150
Epoch 4/8
200/200 [==============================] - 19s 93ms/step - loss: 0.2290 - accuracy: 0.9172 - val_loss: 0.5089 - val_accuracy: 0.8188
Epoch 5/8
200/200 [==============================] - 19s 95ms/step - loss: 0.1673 - accuracy: 0.9395 - val_loss: 0.5919 - val_accuracy: 0.8200
Epoch 6/8
200/200 [==============================] - 19s 94ms/step - loss: 0.1138 - accuracy: 0.9616 - val_loss: 0.6572 - val_accuracy: 0.8037
Epoch 7/8
200/200 [==============================] - 19s 93ms/step - loss: 0.0933 - accuracy: 0.9716 - val_loss: 0.7611 - val_accuracy: 0.8037

In [29]:
pred=model.predict(x_test)

63/63 [==============================] - 2s 30ms/step


In [30]:
pred

array([[4.5968805e-05, 4.2370101e-03, 9.9571699e-01],
       [9.3290641e-05, 1.1770407e-03, 9.9872965e-01],
       [2.2219445e-04, 5.6140567e-04, 9.9921644e-01],
       ...,
       [2.6612438e-04, 2.9307411e-03, 9.9680316e-01],
       [5.5941349e-05, 1.1825296e-03, 9.9876153e-01],
       [1.9447938e-04, 7.8670267e-04, 9.9901879e-01]], dtype=float32)

In [31]:
pred_f=pred.argmax(axis=1)
pred_f

array([2, 2, 2, ..., 2, 2, 2], dtype=int64)

In [32]:
from sklearn.metrics import *

In [33]:
accuracy_LSTM=accuracy_score(y_test,pred_f)*100
accuracy_LSTM

80.2

In [34]:
print(classification_report(y_test,pred_f))

              precision    recall  f1-score   support

           0       0.57      0.57      0.57       212
           1       0.34      0.31      0.32       237
           2       0.90      0.91      0.90      1551

    accuracy                           0.80      2000
   macro avg       0.60      0.60      0.60      2000
weighted avg       0.80      0.80      0.80      2000



# Bidirectional LSTM

In [79]:
import numpy as np
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, LSTM, Dense,Bidirectional


In [80]:
data=pd.read_csv(r"C:\Users\PRANAV\Downloads\archive (16)\Datafiniti_Hotel_Reviews_Jun19.csv")
data.head()

id             dateAdded           dateUpdated  \
0  AWE2FvX5RxPSIh2RscTK  2018-01-18T18:43:12Z  2019-05-20T23:55:47Z   
1  AVwcj_OhkufWRAb5wi9T  2016-11-06T20:21:05Z  2019-05-20T23:31:56Z   
2  AVwcj_OhkufWRAb5wi9T  2016-11-06T20:21:05Z  2019-05-20T23:31:56Z   
3  AVwcj_OhkufWRAb5wi9T  2016-11-06T20:21:05Z  2019-05-20T23:31:56Z   
4  AVwcj_OhkufWRAb5wi9T  2016-11-06T20:21:05Z  2019-05-20T23:31:56Z   

                          address  \
0                 5620 Calle Real   
1  5th And San Carlos PO Box 3574   
2  5th And San Carlos PO Box 3574   
3  5th And San Carlos PO Box 3574   
4  5th And San Carlos PO Box 3574   

                                          categories  \
0  Hotels,Hotels and motels,Hotel and motel mgmt....   
1                               Hotels,Lodging,Hotel   
2                               Hotels,Lodging,Hotel   
3                               Hotels,Lodging,Hotel   
4                               Hotels,Lodging,Hotel   

               primaryCategories               city country  \
0  Accommodation & Food Services             Goleta      US   
1  Accommodation & Food Services  Carmel by the Sea      US   
2  Accommodation & Food Services  Carmel by the Sea      US   
3  Accommodation & Food Services  Carmel by the Sea      US   
4  Accommodation & Food Services  Carmel by the Sea      US   

                                                keys  latitude  ...  \
0             us/ca/goleta/5620callereal/-1127060008  34.44178  ...   
1  us/ca/carmelbythesea/5thandsancarlospobox3574/...  36.55722  ...   
2  us/ca/carmelbythesea/5thandsancarlospobox3574/...  36.55722  ...   
3  us/ca/carmelbythesea/5thandsancarlospobox3574/...  36.55722  ...   
4  us/ca/carmelbythesea/5thandsancarlospobox3574/...  36.55722  ...   

                            reviews.dateSeen reviews.rating  \
0                       2018-01-03T00:00:00Z              3   
1                       2016-10-09T00:00:00Z              4   
2                       2016-10-09T00:00:00Z              3   
3                       2016-10-31T00:00:00Z              5   
4  2016-10-09T00:00:00Z,2016-03-27T00:00:00Z              2   

                                  reviews.sourceURLs  \
0  https://www.tripadvisor.com/Hotel_Review-g3243...   
1  https://www.tripadvisor.com/Hotel_Review-g3217...   
2  https://www.tripadvisor.com/Hotel_Review-g3217...   
3  https://www.tripadvisor.com/Hotel_Review-g3217...   
4  https://www.tripadvisor.com/Hotel_Review-g3217...   

                                        reviews.text  \
0  This hotel was nice and quiet. Did not know, t...   
1  We stayed in the king suite with the separatio...   
2  Parking was horrible, somebody ran into my ren...   
3  Not cheap but excellent location. Price is som...   
4  If you get the room that they advertised on th...   

                                       reviews.title  reviews.userCity  \
0                            Best Western Plus Hotel          San Jose   
1  Clean rooms at solid rates in the heart of Carmel     San Francisco   
2                                           Business   Prescott Valley   
3                                          Very good          Guaynabo   
4                       Low chance to come back here              Reno   

  reviews.userProvince  reviews.username  \
0         UnitedStates      tatsurok2018   
1                   CA         STEPHEN N   
2                   AZ         15Deborah   
3                   PR        Wilfredo M   
4                   NV             Luc D   

                                          sourceURLs  \
0  https://www.tripadvisor.com/Hotel_Review-g3243...   
1  http://www.tripadvisor.com/Hotel_Review-g32172...   
2  http://www.tripadvisor.com/Hotel_Review-g32172...   
3  http://www.tripadvisor.com/Hotel_Review-g32172...   
4  http://www.tripadvisor.com/Hotel_Review-g32172...   

                                            websites  
0  https://www.bestwestern.com/en_US/book/hotel-r...  
1                 

In [102]:

data=data.loc[::,['reviews.rating','reviews.text']]
data

reviews.rating                                       reviews.text
0                  3  this hotel was nice and quiet. did not know, t...
1                  4  we stayed in the king suite with the separatio...
2                  3  parking was horrible, somebody ran into my ren...
3                  5  not cheap but excellent location. price is som...
4                  2  if you get the room that they advertised on th...
...              ...                                                ...
9995               4  my friends and i took a trip to hampton for th...
9996               5  from check in to departure, staff is friendly,...
9997               5  this hampton is located on a quiet street acro...
9998               5  awesome wings (my favorite was garlic parmesan...
9999               4  clean facility just off freeway ..... staff fr...

[10000 rows x 2 columns]

In [103]:
data['reviews.text'].apply(text_process)

0       [hotel, nice, quiet, know, train, track, near,...
1       [stayed, king, suite, separation, bedroom, liv...
2       [parking, horrible, somebody, ran, rental, car...
3       [cheap, excellent, location, price, somewhat, ...
4       [get, room, advertised, website, paid, may, lu...
                              ...                        
9995    [friends, took, trip, hampton, basslights, sho...
9996    [check, departure, staff, friendly, profession...
9997    [hampton, located, quiet, street, across, hosp...
9998    [awesome, wings, favorite, garlic, parmesan, g...
9999    [clean, facility, freeway, staff, friendly, ef...
Name: reviews.text, Length: 10000, dtype: object

In [104]:

data['reviews.rating'].replace({1:0,2:0,3:1,4:2,5:2},inplace=True) 

In [105]:
data_x=data.iloc[:,1]
data_y=data.iloc[:,0]
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(data_x,data_y,test_size=0.2)

In [106]:

y_train=to_categorical(y_train)

In [107]:
max_num_words=10000        # max words to be taken from entire corpus
seq_len=150                 # words taken from each document
embedding_size=200          # embedding size of each word


In [108]:
from keras.preprocessing.text import Tokenizer # is used for tokenization assign id(num) to each word
from keras.preprocessing.sequence import pad_sequences # used to bring all samples to same length by adding extra 0

In [109]:

tokenizer=Tokenizer(num_words=max_num_words)
tokenizer.fit_on_texts(data['reviews.text'])
x_train=tokenizer.texts_to_sequences(x_train)        # will convert text to sequences of Ids
x_train=pad_sequences(x_train,maxlen=seq_len)


In [110]:

x_test=tokenizer.texts_to_sequences(x_test)
x_test=pad_sequences(x_test,maxlen=seq_len)

In [111]:
model=Sequential()  # initilize the network
model.add(Embedding(input_dim=max_num_words,
                   input_length=seq_len,
                   output_dim=embedding_size))

In [112]:

model.add(Bidirectional(LSTM(100)))
model.add(Dense(3,activation='softmax'))  # this 3 bcoz u have 3 level in target variable
from tensorflow.keras.optimizers import Adam
adam=Adam(lr=0.001)
model.compile(optimizer='adam',loss='categorical_crossentropy',metrics=['accuracy'])

In [113]:
model.fit(x_train,y_train,epochs=8,batch_size=32,validation_split=.2)

Epoch 1/8
200/200 [==============================] - 46s 214ms/step - loss: 0.5955 - accuracy: 0.7884 - val_loss: 0.4604 - val_accuracy: 0.8219
Epoch 2/8
200/200 [==============================] - 45s 227ms/step - loss: 0.4208 - accuracy: 0.8364 - val_loss: 0.4324 - val_accuracy: 0.8300
Epoch 3/8
200/200 [==============================] - 45s 223ms/step - loss: 0.3085 - accuracy: 0.8753 - val_loss: 0.4635 - val_accuracy: 0.8200
Epoch 4/8
200/200 [==============================] - 45s 223ms/step - loss: 0.2207 - accuracy: 0.9181 - val_loss: 0.5105 - val_accuracy: 0.8325
Epoch 5/8
200/200 [==============================] - 47s 236ms/step - loss: 0.1467 - accuracy: 0.9484 - val_loss: 0.5905 - val_accuracy: 0.8163
Epoch 6/8
200/200 [==============================] - 46s 228ms/step - loss: 0.0971 - accuracy: 0.9652 - val_loss: 0.7457 - val_accuracy: 0.7544
Epoch 7/8
200/200 [==============================] - 45s 225ms/step - loss: 0.0715 - accuracy: 0.9744 - val_loss: 0.7531 - val_accuracy:

In [114]:
pred=model.predict(x_test)

63/63 [==============================] - 3s 37ms/step


In [115]:
pred_f=pred.argmax(axis=1)
pred_f

array([2, 2, 2, ..., 2, 2, 2], dtype=int64)

In [116]:
from sklearn.metrics import *

In [117]:
accuracy_Bidirectional_LSTM=accuracy_score(y_test,pred_f)*100
accuracy_Bidirectional_LSTM

80.15

In [118]:
print(classification_report(y_test,pred_f))

              precision    recall  f1-score   support

           0       0.60      0.64      0.62       230
           1       0.29      0.28      0.28       236
           2       0.91      0.91      0.91      1534

    accuracy                           0.80      2000
   macro avg       0.60      0.61      0.60      2000
weighted avg       0.80      0.80      0.80      2000



# OPTIMUM MODEL

SO AFTER COMPARING ALL THE MODELS WE ARE GETTING OPTIMUM MODEL IN NAIYE BAYES






accuracy_tdm= 80.15%                               # TDM



accuracy_tf_idf= 79.25%                          # TF-IDF



accuracy_Naiye_Bayes= 85%                     # NAIYE_BAYES



accuracy_LSTM= 80.2%                             # LSTM



accuracy_Bidirectional_LSTM= 80.1%                 # BIDIRECTIONAL-LSTM



In [437]:
def predict(text):
    text = text.lower()
    abcd = stopwords.words('english')
    nopunc = [char for char in text if char not in string.punctuation]
    for i in string.punctuation:
        text = text.replace(str(i), " ")
    prediction = ml.predict(bow.transform([text]))[0]
    print('Ratings got is',prediction)

In [454]:
data['reviews.text'][9]

'hello, i have traveled a lot and abroad and by far this is the worst place i have ever booked. i paid got the key and walked in. omg the place is horrible. this place is one of those hotels they show in horror movies where they kill people. i kid you not! this place is horrible. i only stayed... more'

In [455]:
predict('hello, i have traveled a lot and abroad and by far this is the worst place i have ever booked. i paid got the key and walked in. omg the place is horrible. this place is one of those hotels they show in horror movies where they kill people. i kid you not! this place is horrible. i only stayed... more')

Ratings got is 1


# Conclusion

THIS SENTIMENT ANALYSIS  ON HOTEL REVIEWS WILL HELP PROVIDE VALUABLE INSIGHTS TO BUSINESSES IN THE HOSPITALITY INDUSTRY.IN FOLLOWING WAYS:



1>  IT CAN HIGHLIGHT SPECIFIC AREAS WHERE CUSTOMERS ARE DISSATISFIED, SUCH AS POOR SERVICE,CLEANLINESS,ISSUES OR OUTDATES FACILITIES.




2>  BY ANALYZING SENTIMENTS BUSINESSES CAN PRIORITIZE IMPROVEMENTS BASED ON WHAT MATTERS MOST TO CUSTOMERS.THIS CAN LEAD TO TARGETED INVESTMENTS IN AREAS THAT HAVE THE MOST SIGNIFICANT IMPACT ON CUSTOMER SATISFACTION





3> POSITIVE SENTIMENTS EXPRESSED BY GUESTS CAN BE USED IN MARKETING MATERIALS TO SHOWCASE THE STRENGTHS AND UNIQUE SELLING POINTS OF THE HOTEL. THIS POSITIVE REINFORCEMENT CAN ATTRACT NEW CUSTOMERS.

    
    
 

